src/us/kbase/narrativejobservice/db/ExecEngineMongoDb.java

```diff
@@ -1,240 +1,268 @@
+			for (final Object obj: (BasicBSONList) dbo) {
+				ret.add(cleanObject(obj));
+			}
+		} else if (dbo instanceof BSONObject) {
+			// can't stream because streams don't like null values at HashMap.merge()
+			final BSONObject m = (BSONObject) dbo;
+			final Map<String, Object> ret = new HashMap<>();
+			for (final String k: m.keySet()) {
+				if (!k.equals("_id")) {
+					final Object v = m.get(k);
+					ret.put(k, cleanObject(v));
+				}
+			}
+			return ret;
+		}
+		return dbo;
+	}
 
+	public String[] getSubJobIds(String ujsJobId) throws Exception{
+		// there should be a null/empty check for the ujs id here
+		final DBCursor dbc = taskCol.find(
+				new BasicDBObject("parent_job_id", ujsJobId),
+				new BasicDBObject("ujs_job_id", 1));
 		List<String> idList = new ArrayList<String>();
-		while (ids.hasNext()) {
-			idList.add(ids.next().getUjsJobId());
+		for (final DBObject dbo: dbc) {
+			idList.add((String) dbo.get("ujs_job_id"));
 		}
 		return idList.toArray(new String[idList.size()]);
 	}
 
-
 	public ExecLog getExecLog(String ujsJobId) throws Exception {
-		List<ExecLog> ret = Lists.newArrayList(execLogs.find(
-				String.format("{%s:#}", PK_EXEC_LOGS), ujsJobId)
-				.projection(String.format("{%s:1,%s:1,%s:1}", PK_EXEC_LOGS,
-						"original_line_count", "stored_line_count")).as(ExecLog.class));
-		return ret.size() > 0 ? ret.get(0) : null;
+		// there should be a null/empty check for the ujs id here
+		// should make these strings constants
+		final DBObject log = logCol.findOne(
+				new BasicDBObject(PK_EXEC_LOGS, ujsJobId),
+				new BasicDBObject(PK_EXEC_LOGS, 1)
+						.append("original_line_count", 1)
+						.append("stored_line_count", 1));
+		final ExecLog ret;
+		if (log == null) {
+			ret = null;
+		} else {
+			ret = new ExecLog();
+			ret.setOriginalLineCount((Integer) log.get("original_line_count"));
+			ret.setStoredLineCount((Integer) log.get("stored_line_count"));
+			ret.setUjsJobId((String) log.get(PK_EXEC_LOGS));
+		}
+		return ret;
 	}
 
 	public void insertExecLog(ExecLog execLog) throws Exception {
-		execLogs.insert(execLog);
+		// should be a null check here
+		insertExecLogs(Arrays.asList(execLog));
 	}
 
 	public void insertExecLogs(List<ExecLog> execLogList) throws Exception {
-		Object[] execLogArray = execLogList.toArray(new Object[execLogList.size()]);
-		execLogs.insert(execLogArray);
+		// should be a null collection contents check here
+		logCol.insert(execLogList.stream().map(l -> toDBObj(l)).collect(Collectors.toList()));
 	}
 
 	public void updateExecLogLines(String ujsJobId, int newLineCount,
 								   List<ExecLogLine> newLines) throws Exception {
-		execLogs.update(String.format("{%s:#}", PK_EXEC_LOGS), ujsJobId).with(
-				String.format("{$set:{%s:#,%s:#},$push:{%s:{$each:#}}}",
-						"original_line_count", "stored_line_count", "lines"),
-				newLineCount, newLineCount, newLines);
+		// needs input checking
+		logCol.update(new BasicDBObject(PK_EXEC_LOGS, ujsJobId),
```

Comment:
this section is a little bit of a nightmare to look at. Can it be split up for readability?

Reply:
Is
```
		BasicDBObject ujsJobIdQuery = new BasicDBObject(PK_EXEC_LOGS, ujsJobId);

		BasicDBObject each = new BasicDBObject("lines", new BasicDBObject("$each",
				newLines.stream().map(l -> toDBObj(l))
						.collect(Collectors.toList())));

		BasicDBObject set = new BasicDBObject("$set", new BasicDBObject("original_line_count", newLineCount).append("stored_line_count", newLineCount))
				.append("$push", each);

		logCol.update(ujsJobIdQuery,set);
```
Better or worse?

Reply:
For me I think it's better.  Maybe @MrCreosote would have an opinion on it as well.

Reply:
lol no